core/src/main/java/com/sun/mail/util/SocketFetcher.java

```diff
@@ -430,6 +432,54 @@ private static SocketFactory getSocketFactory(String sfClass)
 	return sf;
     }
 
+	/**
+	 * Return an instance of {@link HostnameVerifier}.
+	 * This method assumes the {@link HostnameVerifier} class provides an
+	 * accessible default constructor to instantiate the instance.
+	 *
+	 * @param hnvClassname               the class name of the {@link HostnameVerifier}.
+	 * @return                           the {@link HostnameVerifier}
+	 * @throws ClassNotFoundException    If the {@link HostnameVerifier} class cannot be found in the current class loader after also not being found in the context class loader.
+	 * @throws NoSuchMethodException     If the {@link HostnameVerifier} class does not implement a publicly accessible default constructor.
+	 * @throws InvocationTargetException
+	 * @throws InstantiationException
+	 * @throws IllegalAccessException
+	 */
+	private static HostnameVerifier getHostnameVerifier(String hnvClassname)
+			throws ClassNotFoundException,
+			NoSuchMethodException,
+			InvocationTargetException,
+			InstantiationException,
+			IllegalAccessException {
+		if (!(hnvClassname == null || hnvClassname.length() == 0)) {
+			ClassLoader ccl = getContextClassLoader();
+			Class<?> clsHostnameVerifier = null;
+
+			// Attempt to load the class from the context class loader.
+			if (ccl != null) {
+				try {
+					clsHostnameVerifier = Class.forName(hnvClassname, false, ccl);
+				} catch (ClassNotFoundException cex) {
+					// Ignore it - try the current class loader
+				}
+			}
+
+			// If the class was not resolved in the context class loader, try the current class loader.
+			if (clsHostnameVerifier == null) {
+				clsHostnameVerifier = Class.forName(hnvClassname);
+			}
+
+			// If we were able to resolve the class object, attempt to construct an instance
+			if (clsHostnameVerifier != null) {
+				// Attempt to invoke the default constructor
```

Comment:
We should avoid invoking the constructor if the type is wrong as there would be no general contract as to what this unknown class is going to do when we create it.  We do these checks in the j.m.Session when we create services.

Check that the clsHostnameVerifier class can be assigned to javax.net.ssl.HostnameVerifier.  If it can't then fail with a class cast exception with a nice message.